rfcs/rfc-111.md

```diff
@@ -0,0 +1,354 @@
+    }
+}
+```
+
+This is a divergence from the initial implementation, as users would have to specify "Properties" in a path before checking properties, but this allows for a more standardized way to check resource properties; authors no longer need to use the dot prefix to access keys on this level like DeletionPolicy and Metadata.
+
+E.g.  taking a modified example from verbosity issue we described above
+```
+AWS::ApiGateway::Method AuthorizationType == NONE
+AWS::ApiGateway::Method Integration == {"IntegrationHttpMethod":"POST","Type":"AWS_PROXY","Uri":"arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${LambdaWAFBadBotParserFunction.Arn}/invocations"}
+AWS::ApiGateway::Method ResourceId == ApiGatewayBadBot.RootResourceId |OR| AWS::ApiGateway::Method ResourceId == ApiGatewayBadBotResource
+AWS::ApiGateway::Method .DeletionPolicy == Retain
+AWS::ApiGateway::Method .Metadata.CreatedBy == SecureCFNGenerationTool
+```
+
+The ruleset above converted to use a type block:
+```
+AWS::ApiGateway::Method {
+    Properties.AuthorizationType == "NONE"
+    Properties.Integration == {"IntegrationHttpMethod":"POST","Type":"AWS_PROXY","Uri":"arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${LambdaWAFBadBotParserFunction.Arn}/invocations"}
+    Properties.ResourceId == "ApiGatewayBadBot.RootResourceId" or
+    Properties.ResourceId == "ApiGatewayBadBotResource"
+    DeletionPolicy == "Retain"
+    Metadata.CreatedBy == "SecureCFNGenerationTool"
+}
+```
+Customers can continue to use the simplicity of expression from the previous language without the repetition of the type name.
+
+### **Named-Rules**
+
+A named rule is a collection of clauses in the CNF form. Named rules allow for expressing powerful combinations to provide more complex evaluation than primitive rules. Named rules allow for re-use and improved composition and remove verbosity and repetition. This also allows rule evaluators to concisely identify passing/failing rules in the evaluation output. Named rules take the following form:
+
+<pre>
+[<b>rule</b> <em>rule_name</em>] [<b>when</b> <em>conditions</em>] {
+    <em>clauses</em>
+    <em>type-blocks</em>
+    <em>assignments</em>
+}
+</pre>
+
+Here, _rule_ keyword designates the start of a named rule block. The keyword is followed by the *rule_name* that is a human readable name. Rule names are optional but provide context in output as to what exactly failed. When evaluating the rules file, the *rule_name* is displayed along with with status for the evaluation <b>PASS. FAIL or SKIP</b>. The rule name can be followed by optional conditions (_When_ guards) that act as a guard to determine if the rule is application for evaluation or must be skipped, a.k.a conditionally evaluated (akin to WHEN CHECK in the previous language). 
+
+
+The block contains a set of clauses in Conjunctive Normal Form. To simplify authoring clauses and provide a consistent interpretation model, the following rules apply:
```

Comment:
I'd rather see the rule syntax support parenthesis. The CNF form, the explicit OR vs implicit AND etc are new rules that developers have to learn. Yes it can be abused to write complex rules. But that should be caught by normal s/w sdlc, code reviews etc. I personally prefer this construction. 

```
rule example_rule when stage == 'prod' {
        let ec2_instance_types := [/^t*/, /^m*/]  
	
	dependent_rule
	AND
        AWS::EC2::Instance InstanceType IN %ec2_instance_types 
	AND 
	(
		AWS::EC2::Instance {                         
			xxxxx 
		}
		OR		
		AWS::EC2::Instance {        
                        yyyyy    
    		}
	)
}
```

Reply:
so with this construction, you would have to specify AND and couldn't implicitly rely on the next rule being ANDed, right? I think as long as it is explicit blocks being ANDed makes sense. this seems a repetitive, but with the added benefit of being explicit as possible.